packages/shadycss/shadow-parts.md

```diff
@@ -0,0 +1,307 @@
+
+```css
+x-a x-a.odd[shady-part~=x-a:x-a:num] {
+  color: red;
+}
+x-a x-a.even[shady-part~=x-a:x-a:num] {
+  color: green;
+}
+```
+
+Workarounds will vary based on specific scenario. For the above example, a good
+workaround would be to use two unique part names: `numOdd`, `numEven`.
+
+## How it works
+
+The basic approach is to convert `::part` rules to attribute selectors that
+encode the _providing_ scope, the _receiving_ scope, and the part name. Each
+shadow host then discovers its parent scope, and applies the correct attribute
+to each of its part nodes.
+
+Native:
+
+```html
+<x-a>
+  #shadow-root
+    <style>
+      x-b::part(greeting) {
+        color: red;
+      }
+    </style>
+    <x-b>
+      #shadow-root
+        <div part="greeting">Hello</div>
+    </x-b>
+</x-a>
+```
+
+ShadyCSS:
+
+```html
+<style>
+  x-a x-b [shady-part~=x-a:x-b:greeting] {
+    color: red;
+  }
+</style>
+
+<x-a>
+  <x-b>
+    <div shady-part="x-a:x-b:greeting">Hello</div>
+  </x-b>
+</x-a>
+```
+
+### Document styles
+
+As is standard in ShadyCSS, document-level styles must be opted-in with
+`ShadyCSS.CustomStyleInterface.addCustomStyle`. The first component of a shimmed
+document-level `::part` style rule will be literally `document`:
+
+```html
+<div shady-part="document:x-b:greeting">
+```
+
+### Multiple parts
+
+When a `::part` rule lists multiple space-separated part names, then only part
+nodes with all of those part names match. This is emulated with:
+
+Native:
+
+```css
+x-b::part(foo bar) {
+  color: red;
+}
+```
+
+ShadyCSS:
+
+```css
+x-a x-b [shady-part~=x-a:x-b:foo][shady-part~=x-a:x-b:bar] {
+  color: red;
+}
+```
+
+### Exportparts
+
+In the basic form, `::part` rules only flow down one hop in the scope tree, to a
+shadow root's children. The `exportparts` attribute is used to propagate styles
+more deeply.
+
+ShadyCSS emulates `exportparts` by walking up the DOM for each instance (when
+`styleElement` is called), following `exportparts` attributes. This is cached,
+so each shadow root only needs to be walked once even there are multiple parts
+below it. For each relevant scope, an additional `shady-part` attribute is added
+to part nodes.
+
+Native:
+
+```html
+<x-a>
+  #shadow-root
+    <style>
+      x-b::part(salutation) {
+        color: red;
+      }
+    </style>
+    <x-b>
+      #shadow-root
+        <style>
+          x-c::part(greeting) {
+            font-weight: bold;
+          }
+        </style>
+        <x-c exportparts="greeting:salutation">
+          #shadow-root
+            <div part="greeting">Hello</div>
+        </x-c>
+    </x-b>
+</x-a>
+```
+
+ShadyCSS:
+
+```html
+<style>
+  x-a x-b [shady-part~=x-a:x-b:salutation] {
+    color: red;
+  }
+  x-b x-c [shady-part~=x-b:x-c:greeting] {
+    font-weight: bold;
+  }
+</style>
+
+<x-a>
+  <x-b>
+    <x-c>
+      <div shady-part="x-b:x-c:greeting
+                       x-a:x-b:salutation">Hello</div>
+    </x-c>
+  </x-b>
+</x-a>
+```
+
+### Custom properties
```

Comment:
Clarify that this is only needed on browsers that don't natively support css custom properties.